File: kernel/mri/mri.c
/*
 * Copyright (c) 2002, 2017 Jens Keiner, Stefan Kunis, Daniel Potts
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 51
 * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include "config.h"

#include <string.h>
#include <math.h>
#ifdef HAVE_COMPLEX_H
#include <complex.h>
#endif
#include "nfft3.h"
#include "infft.h"

/**
 * window_funct_plan is a plan to use the window functions
 * independent of the nfft
 */

typedef struct window_funct_plan_ {
  int d;
	int m;
	int n[1];
	double sigma[1];
	double *b;
} window_funct_plan;

/**
 * init the window_funct_plan
 */
static void window_funct_init(window_funct_plan* ths, int m, int n, double sigma) {
	ths->d=1;
	ths->m=m;
	ths->n[0]=n;
	ths->sigma[0]=sigma;
  WINDOW_HELP_INIT
}

/*
 * mri_inh_2d1d
 */

void mri_inh_2d1d_trafo(mri_inh_2d1d_plan *that) {
  int l,j;
  double _Complex *f = (double _Complex*) nfft_malloc(that->M_total*sizeof(double _Complex));
  double _Complex *f_hat = (double _Complex*) nfft_malloc(that->N_total*sizeof(double _Complex));

  window_funct_plan *ths = (window_funct_plan*) nfft_malloc(sizeof(window_funct_plan));
	window_funct_init(ths,that->plan.m,that->N3,that->sigma3);

	/* the pointers that->f and that->f_hat have been modified by the solver */
	that->plan.f = that->f;
  that->plan.f_hat = that->f_hat;


	memset(f,0,that->M_total*sizeof(double _Complex));
  for(j=0;j<that->N_total;j++)
  {
    f_hat[j]=that->f_hat[j];
  }

  for(l=-ths->n[0]/2;l<=ths->n[0]/2;l++) {
    for(j=0;j<that->N_total;j++)
      that->f_hat[j]*=cexp(-2*KPI*_Complex_I*that->w[j]*((double)l))/PHI_HUT(ths->n[0], ths->n[0]*that->w[j],0);
    nfft_trafo(&that->plan);
    for(j=0;j<that->M_total;j++){
      /* PHI has compact support */
			if(fabs(that->t[j]-((double)l)/((double)ths->n[0]))<that->plan.m/((double)ths->n[0]))
      {
        double phi_val = PHI(ths->n[0],that->t[j]-((double)l)/((double)ths->n[0]),0);
        f[j]+=that->f[j]*phi_val;
// the line below causes internal compiler error for gcc 4.7.1
//        f[j]+=that->f[j]*PHI(ths->n[0],that->t[j]-((double)l)/((double)ths->n[0]),0);
      }
    }
    for(j=0;j<that->N_total;j++)
      that->f_hat[j]=f_hat[j];
  }

  nfft_free(that->plan.f);
  that->f=f;
  that->plan.f = that->f;

  nfft_free(f_hat);

  WINDOW_HELP_FINALIZE
  nfft_free(ths);
}

void mri_inh_2d1d_adjoint(mri_inh_2d1d_plan *that) {
  int l,j;
  double _Complex *f = (double _Complex*) nfft_malloc(that->M_total*sizeof(double _Complex));
  double _Complex *f_hat = (double _Complex*) nfft_malloc(that->N_total*sizeof(double _Complex));

  window_funct_plan *ths = (window_funct_plan*) nfft_malloc(sizeof(window_funct_plan));
	window_funct_init(ths,that->plan.m,that->N3,that->sigma3);

	memset(f_hat,0,that->N_total*sizeof(double _Complex));

	/* the pointers that->f and that->f_hat have been modified by the solver */
	that->plan.f = that->f;
  that->plan.f_hat = that->f_hat;

	for(j=0;j<that->M_total;j++)
  {
    f[j]=that->f[j];
  }



  for(l=-ths->n[0]/2;l<=ths->n[0]/2;l++) {

    for(j=0;j<that->M_total;j++) {
      /* PHI has compact support */
      if(fabs(that->t[j]-((double)l)/((double)ths->n[0]))<that->plan.m/((double)ths->n[0]))
        that->f[j]*=PHI(ths->n[0],that->t[j]-((double)l)/((double)ths->n[0]),0);
      else
      	that->f[j]=0.0;
    }
    nfft_adjoint(&that->plan);
    for(j=0;j<that->N_total;j++)
      f_hat[j]+=that->f_hat[j]*cexp(2*KPI*_Complex_I*that->w[j]*((double)l));
    for(j=0;j<that->M_total;j++)
      that->f[j]=f[j];
  }

  for(j=0;j<that->N_total;j++)
  {
    f_hat[j] /= PHI_HUT(ths->n[0],ths->n[0]*that->w[j],0);
  }

  nfft_free(that->plan.f_hat);
  that->f_hat=f_hat;
  that->plan.f_hat = that->f_hat;

	nfft_free(f);

	WINDOW_HELP_FINALIZE
  nfft_free(ths);
}

void mri_inh_2d1d_init_guru(mri_inh_2d1d_plan *ths, int *N, int M, int *n,
                    int m, double sigma, unsigned nfft_flags, unsigned fftw_flags) {

  nfft_init_guru(&ths->plan,2,N,M,n,m,nfft_flags,fftw_flags);
  ths->N3=N[2];
	ths->sigma3=sigma;
  ths->N_total = ths->plan.N_total;
  ths->M_total = ths->plan.M_total;
  ths->f = ths->plan.f;
  ths->f_hat = ths->plan.f_hat;

  ths->t = (double*) nfft_malloc(ths->M_total*sizeof(double));
  ths->w = (double*) nfft_malloc(ths->N_total*sizeof(double));

  ths->mv_trafo = (void (*) (void* ))mri_inh_2d1d_trafo;
  ths->mv_adjoint = (void (*) (void* ))mri_inh_2d1d_adjoint;
}

void mri_inh_2d1d_finalize(mri_inh_2d1d_plan *ths) {
  nfft_free(ths->t);
  nfft_free(ths->w);

	/* the pointers ths->f and ths->f_hat have been modified by the solver */
	ths->plan.f = ths->f;
  ths->plan.f_hat = ths->f_hat;

	nfft_finalize(&ths->plan);
}

/*
 * mri_inh_3d
 */

void mri_inh_3d_trafo(mri_inh_3d_plan *that) {
  int l,j;
  window_funct_plan *ths = (window_funct_plan*) nfft_malloc(sizeof(window_funct_plan));
	window_funct_init(ths,that->plan.m,that->N3,that->sigma3);

	/* the pointers that->f has been modified by the solver */
  that->plan.f =that->f ;



  for(j=0;j<that->N_total;j++) {
    for(l=-ths->n[0]/2;l<ths->n[0]/2;l++)
    {
      /* PHI has compact support */
      if(fabs(that->w[j]-((double)l)/((double)ths->n[0]))<ths->m/((double)ths->n[0]))
        that->plan.f_hat[j*ths->n[0]+(l+ths->n[0]/2)]= that->f_hat[j]*PHI(ths->n[0],that->w[j]-((double)l)/((double)ths->n[0]),0);
      else
	      that->plan.f_hat[j*ths->n[0]+(l+ths->n[0]/2)]=0.0;
    }
  }

  nfft_trafo(&that->plan);

  for(j=0;j<that->M_total;j++)
  {
    that->f[j] /= PHI_HUT(ths->n[0],ths->n[0]*that->plan.x[3*j+2],0);
  }

	WINDOW_HELP_FINALIZE
  nfft_free(ths);
}

void mri_inh_3d_adjoint(mri_inh_3d_plan *that) {
  int l,j;
  window_funct_plan *ths = (window_funct_plan*) nfft_malloc(sizeof(window_funct_plan));
	window_funct_init(ths,that->plan.m,that->N3,that->sigma3);

	/* the pointers that->f has been modified by the solver */
  that->plan.f =that->f ;

  for(j=0;j<that->M_total;j++)
  {
    that->f[j] /= PHI_HUT(ths->n[0],ths->n[0]*that->plan.x[3*j+2],0);
  }

  nfft_adjoint(&that->plan);

  for(j=0;j<that->N_total;j++) {
    that->f_hat[j]=0.0;
    for(l=-ths->n[0]/2;l<ths->n[0]/2;l++)
    {
      /* PHI has compact support */
      if(fabs(that->w[j]-((double)l)/((double)ths->n[0]))<ths->m/((double)ths->n[0]))
        that->f_hat[j]+= that->plan.f_hat[j*ths->n[0]+(l+ths->n[0]/2)]*PHI(ths->n[0],that->w[j]-((double)l)/((double)ths->n[0]),0);
    }
  }


	WINDOW_HELP_FINALIZE
  nfft_free(ths);
}

void mri_inh_3d_init_guru(mri_inh_3d_plan *ths, int *N, int M, int *n,
                    int m, double sigma, unsigned nfft_flags, unsigned fftw_flags) {
  ths->N3=N[2];
	ths->sigma3=sigma;
  nfft_init_guru(&ths->plan,3,N,M,n,m,nfft_flags,fftw_flags);
  ths->N_total = N[0]*N[1];
  ths->M_total = ths->plan.M_total;
  ths->f = ths->plan.f;
  ths->f_hat = (double _Complex*) nfft_malloc(ths->N_total*sizeof(double _Complex));
  ths->w = (double*) nfft_malloc(ths->N_total*sizeof(double));

  ths->mv_trafo = (void (*) (void* ))mri_inh_3d_trafo;
  ths->mv_adjoint = (void (*) (void* ))mri_inh_3d_adjoint;
}

void mri_inh_3d_finalize(mri_inh_3d_plan *ths) {
  nfft_free(ths->w);
  nfft_free(ths->f_hat);
  nfft_finalize(&ths->plan);
}

File: applications/mri/mri3d/mri.m
% Copyright (c) 2002, 2017 Jens Keiner, Stefan Kunis, Daniel Potts
%
% This program is free software; you can redistribute it and/or modify it under
% the terms of the GNU General Public License as published by the Free Software
% Foundation; either version 2 of the License, or (at your option) any later
% version.
%
% This program is distributed in the hope that it will be useful, but WITHOUT
% ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
% FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
% details.
%
% You should have received a copy of the GNU General Public License along with
% this program; if not, write to the Free Software Foundation, Inc., 51
% Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
%
% This script file is an example of the usage

N=48;   % points per row / column
Z=48;    % number of slices

% Construct the raw data of the phantom
% and write it to input_f.dat
% To use another example than the phantom
% just put the reshape of your example into input_f.dat
construct_phantom(N,Z);

% Construct the knots in k-space and write them to knots.dat
% Different knots like spiral,rose,radial,radial_3d or linogram can be chosen
% The radial_3d knots just work with construct_data_3d and reconstruct_data_3d
% Then the weights are generated in construct_knots_radial_3d
% M is the number of knots
M = construct_knots_spiral(N,Z);

% First make N^2 1d-FFT, then Z 2d-NFFT on the constructed knots
% and write the output to output_phantom_nfft.dat
if ispc
    cmd='construct_data_2d1d.exe';
else 
    cmd='./construct_data_2d1d';
end
system([cmd ' output_phantom_nfft.dat '...
         int2str(N) ' ' int2str(M) ' ' int2str(Z)]);

% Precompute the weights using voronoi cells
% and write them to weights.dat
precompute_weights_2d('output_phantom_nfft.dat',M,Z);

% First make Z inverse 2d-NFFT, then N^2 inverse 1d-FFT
% and write the output to output_real.dat and output_imag.dat
% The usage is "./reconstruct_data_2d1d filename N M Z ITER WEIGHTS"
% where ITER is the number of iteration and WEIGHTS is 1
% if the weights are used 0 else
if ispc
    cmd='reconstruct_data_2d1d.exe';
else 
    cmd='./reconstruct_data_2d1d';
end
system([cmd ' output_phantom_nfft.dat ' ...
         int2str(N) ' ' int2str(M) ' ' int2str(Z)  ' 3 1']);

% Visualize the three dimensional phantom. Makes a pic of
% every slice and one plot of the N/2 row of the 10th plane.
[~,~] = mkdir('pics_2+1d');
visualize_data('pics_2+1d/pic', N, Z, 1, 'Inverse 2d1d-NFFT - 3. iteration - spiral knots');

% Compute the root mean square
 rms('pics_2+1d/rms.txt');

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% The same as above but reconstructed with gridding.
% That means first an adjoint 2d-NFFT, then a 1d-FFT.
% The ITER parameter is obsolent and just for compatibility
if ispc
    cmd='reconstruct_data_gridding.exe';
else 
    cmd='./reconstruct_data_gridding';
end
system([cmd ' output_phantom_nfft.dat ' ...
         int2str(N) ' ' int2str(M) ' ' int2str(Z)  ' 0 1']);
[~,~] = mkdir('pics_gridding');
visualize_data('pics_gridding/pic', N, Z, 2, 'Adjoint 2d1d-NFFT (Gridding) - spiral knots');
rms('pics_gridding/rms.txt');

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% The same as above but reconstructed with a 3d-nfft
if ispc
    cmd='reconstruct_data_3d.exe';
else 
    cmd='./reconstruct_data_3d';
end
system([cmd ' output_phantom_nfft.dat ' ...
         int2str(N) ' ' int2str(M) ' ' int2str(Z)  ' 1 1']);
[~,~] = mkdir('pics_3d');
visualize_data('pics_3d/pic', N, Z, 3, 'Inverse 3d-NFFT - 1. iteration - spiral knots');
rms('pics_3d/rms.txt');

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% An example with the radial_3d knots
% please uncomment to try

%M = construct_knots_radial_3d(N,N*3/2);

%system(['./construct_data_3d ' 'output_phantom_nfft.dat '...
%         int2str(N) ' ' int2str(M) ' ' int2str(Z)]);

%system(['./reconstruct_data_3d ' 'output_phantom_nfft.dat ' ...
%         int2str(N) ' ' int2str(M) ' ' int2str(Z)  ' 3 1']);
%visualize_data('pics_3d/pic_', N, Z, 4, 'Inverse 3d-NFFT - 3. iteration - radial_3d knots');
%rms('pics_3d/rms.txt');
